### PR TITLE
Issue #33: convert to package and include typeshed type hints

### PR DESCRIPTION
**Pull request by ncoghlan**

### CODE REVIEW

File: contextlib2/contextlib.pyi [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+_ExitFunc = Callable[[Optional[Type[BaseException]], Optional[BaseException], Optional[TracebackType]], bool]
+_CM_EF = TypeVar("_CM_EF", ContextManager[Any], _ExitFunc)
+
+class _GeneratorContextManager(ContextManager[_T_co]):
+    def __call__(self, func: _F) -> _F: ...
+
+# type ignore to deal with incomplete ParamSpec support in mypy
+def contextmanager(func: Callable[_P, Iterator[_T]]) -> Callable[_P, _GeneratorContextManager[_T]]: ...  # type: ignore
+
+if sys.version_info >= (3, 7):
```
Comment (graingert) [inline contextlib2/contextlib.pyi:41]: Isn't this always defined in contextlib2?

Comment (graingert) [reply contextlib2/contextlib.pyi:41]: You could swap them out for `if True:` to minimize the diff noise?

Comment (ncoghlan) [reply contextlib2/contextlib.pyi:41]: Ah, true. And I like that suggestion for minimising the diff.

Comment (graingert) [reply contextlib2/contextlib.pyi:41]: The ABC will be a bit troublesome as you might need to import them from typing_extensions?

Comment (ncoghlan) [reply contextlib2/contextlib.pyi]: I wasn't sure about that, as the typing module imports were unconditional in the typeshed version as well - it only guarded their aliasing to contextlib API entries.

I figured either mypy (et al) must be doing some shenanigans behind the scenes to let that work even on older Python versions, or else it needed to be fixed in typeshed first, and then contextlib2 could inherit that fix.

Comment (graingert) [reply contextlib2/contextlib.pyi]: Ah you're right, it's already in 3.6 https://docs.python.org/3.6/library/typing.html#typing.AsyncContextManager

Comment (ncoghlan) [reply contextlib2/contextlib.pyi]: Heh, we both found the same info in about the same time, but picked different subthreads to update first :)

**Review by graingert — APPROVED** [review]
Awesome, I'm very excited to use this!

You might want add type annotations for the deprecated classes and methods?

**Comment by ncoghlan** [comment]
There are only two deprecated APIs, and I honestly suspect I should just remove them completely at this point: #31 (not in 21.6.0, though).

I did enhance the CI to actually run mypy, though, and confirmed locally that it failed as expected if there was something wrong in the pyi file.

So I'll merge this now, check everything over again in the morning, and assuming I don't think of anything else I missed, I'll cut the release tomorrow

(And in writing that, I realised I need release notes for this change!)

**Comment by ncoghlan** [comment]
@graingert Ah, that would explain why mypy only complained when I introduced outright errors into the stub, and not when I only made it inconsistent with the actual implementation.

**Pull request merged**